engine/src/test/java/org/hibernate/validator/test/cfg/ValidatorConstraintMappingTest.java

```diff
@@ -0,0 +1,308 @@
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static org.hibernate.validator.testutil.ConstraintViolationAssert.assertCorrectConstraintViolationMessages;
+import static org.hibernate.validator.testutil.ConstraintViolationAssert.assertNumberOfViolations;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javax.validation.Constraint;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import javax.validation.ConstraintViolation;
+import javax.validation.Payload;
+import javax.validation.UnexpectedTypeException;
+import javax.validation.ValidationException;
+import javax.validation.Validator;
+
+import org.hibernate.validator.HibernateValidator;
+import org.hibernate.validator.HibernateValidatorConfiguration;
+import org.hibernate.validator.cfg.ConstraintMapping;
+import org.hibernate.validator.internal.cfg.context.DefaultConstraintMapping;
+import org.hibernate.validator.testutil.TestForIssue;
+import org.hibernate.validator.testutil.ValidatorUtil;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+/**
+ * Unit test for {@link ConstraintMapping#constraint(Class)} et al.
+ *
+ * @author Yoann Rodiere
+ */
+@TestForIssue( jiraKey = "HV-501")
+public class ValidatorConstraintMappingTest {
+
+	private HibernateValidatorConfiguration config;
+	private DefaultConstraintMapping mapping;
+
+	@BeforeMethod
+	public void setUp() {
+		config = ValidatorUtil.getConfiguration( HibernateValidator.class );
+		mapping = (DefaultConstraintMapping) config.createConstraintMapping();
+	}
+
+	@Test(
+			expectedExceptions = IllegalArgumentException.class,
+			expectedExceptionsMessageRegExp = "HV[0-9]*: The annotation type must not be null when creating a constraint definition."
+	)
+	public void testNullClass() {
+		mapping.constraint( null );
+	}
+
+	@Test(
+			expectedExceptions = IllegalArgumentException.class,
+			expectedExceptionsMessageRegExp = "HV[0-9]*: The annotation type must be annotated with @javax.validation.Constraint when creating a constraint definition."
+	)
+	public void testNonConstraintAnnotation() {
+		mapping.constraint( NonConstraintAnnotation.class );
+	}
+
+	@Test
+	public void testConstraintMapping() {
+		mapping.constraint( ConstraintAnnotation.class )
+				.validatedBy( NonDefaultLongValidator.class );
+
+		config.addMapping( mapping );
+		Validator validator = config.buildValidatorFactory().getValidator();
+
+		Set<? extends ConstraintViolation<?>> violations = validator.validate( new ConstrainedLongFieldBean() );
+		assertNumberOfViolations( violations, 1 );
+		assertCorrectValidatorTypes( violations, NonDefaultLongValidator.class );
+	}
+
+	@Test
+	public void testConstraintMappingDefaultsToIncludingExistingValidators() {
```

Comment:
Well, this test mainly ensures that `validatedBy` does not **override** the existing validators, but just add them. So I just assert later that the default validator is still registered after a call to `validatedBy`. This test might pass if the impl was a no-op, but right there that's exactly what we're testing: the absence of side-effect.

As for testing when multiple validators are registered for the same type and the same annotations, `testMultipleValidatorsForSameType` does exactly that. I can add a test without a call to `includeExistingValidators(true)`, if you wish, but since it's the default (and since `testConstraintMappingDefaultsToIncludingExistingValidators` ensures it's the default), I didn't think it would be necessary. Will do anyway, there's no harm in that.